Parallel replicas custom key: skip unavailable replicas

devcrafter opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

devcrafter committed:
Parallel replicas custom key: skip unavailable replicas

devcrafter committed:
Remove commented code

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #57235 from ClickHouse/ClickHouse:pr-custom-key-failover -->
*This is an automated comment for commit 63adf4d20cf86b4261b03665c1c432817a862cdc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57235/63adf4d20cf86b4261b03665c1c432817a862cdc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6998296747">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57235/63adf4d20cf86b4261b03665c1c432817a862cdc/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

devcrafter committed:
Fixes

devcrafter committed:
Fix clang-tidy warnings

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Update test

devcrafter committed:
Change log level to avoid test failure

devcrafter committed:
Fix: use only one connection per replica

devcrafter committed:
Some refactoring and cleanup

+ change load balancing for custom key w/o hedged connections
  (need the test though)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Fix existing intergation test

devcrafter committed:
Remove commented code

devcrafter committed:
Integration test: custom key with failover

robot-clickhouse committed:
Automatic style fix

devcrafter committed:
Cleanup

devcrafter committed:
Fix

devcrafter committed:
Fix race condition

devcrafter committed:
Fix clang-tidy warning: move from moved object

devcrafter committed:
Fix build: GetPriorityForLoadBalancing - explicit copy/move constructor/operator

to make copyable/movable with atomic last_used member

devcrafter commented:
UBSan report - https://s3.amazonaws.com/clickhouse-test-reports/57235/e26e1b691bcc661528e6630147b8627a379337be/fuzzer_astfuzzerubsan/report.html

devcrafter committed:
Fix test

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

Algunenano commented on tests/queries/0_stateless/02918_parallel_replicas_custom_key_unavailable_replica.sql:
```diff
@@ -0,0 +1,33 @@
+SET prefer_localhost_replica=0;
+
+-- { echoOn }
+SELECT y, count()
```
What will happen if we say `max_parallel_replicas=4`? Will the query be `cityHash64(y)/3` or will it be `cityHash64(y)/4` and the missing replica will be reassigned?

devcrafter replied:
It will be `cityHash64(y)/4` and data partitioned by custom key should be reassigned among available replicas in round-robin fashion

Algunenano replied:
That seems wrong to me. It will lead to one replica doing 2x more work.

Can't we instead adapt the condition to  the number of replicas available and disable custom parallel replicas if there is only one?

Algunenano replied:
Now that I think of it, we probably don't want to do that since it involves waiting for a dead/slow replica (which should be the odd case).

Let's just add a test to verify the behaviour and result is correct now

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter commented on tests/config/config.d/clusters.xml:
```diff
@@ -144,6 +144,24 @@
                 </replica>
             </shard>
         </parallel_replicas>
+        <parallel_replicas_custom_key_unavailable_replica>
```
What would be your suggestion? This cluster is for a particular test, so not sure that it's necessary to generalize it. But the general name could be `test_cluster_1_shard_3_replicas_1_unavailable`

Algunenano replied:
Yep, something like that sounds fine

devcrafter committed:
Update tests

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

robot-clickhouse committed:
Automatic style fix

Algunenano commented on tests/integration/test_parallel_replicas_custom_key_failover/test.py:
```diff
@@ -0,0 +1,129 @@
+    )
+
+    # populate data
+    node1.query(
```
What's with the inserts in different replicas here? Are you testing something that I'm missing? It's a ReplicatedMergeTree.

devcrafter replied:
No reason, copy/paste mostly. Will clean up

Algunenano commented on tests/integration/test_parallel_replicas_custom_key_failover/test.py:
```diff
@@ -0,0 +1,129 @@
+    for i in range(4):
+        expected_result += f"{i}\t500\n"
+
+    log_comment = "d304e2a7-ba60-49cd-8bb7-dbb6d9b40952"
```
This query is going to be executed multiple times (with different parameters) so you'll need to randomize the comment if you want to avoid failures (happens in CI too)

Algunenano commented on tests/queries/0_stateless/01361_fover_remote_num_tries.reference:
```diff
@@ -1 +1 @@
-30
+20
```
Does this mean that the number of retries was reduced?

devcrafter committed:
Fix round robin prioirty generation

devcrafter committed:
Fix test: use different log_comment for each test instance

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Fix load balancing for hedged connections

devcrafter committed:
Fix typo

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Fix cluster name

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

Algunenano commented on src/Common/GetPriorityForLoadBalancing.cpp:
```diff
@@ -33,19 +33,17 @@ std::function<Priority(size_t index)> GetPriorityForLoadBalancing::getPriorityFu
-            get_priority = [this, pool_size](size_t i)
+            auto local_last_used = last_used.fetch_add(1) % pool_size;
+
+            get_priority = [pool_size, local_last_used](size_t i)
```
Please consider keeping the comment or replacing it with a more comprehensive one, otherwise it's impossible to understand without taking quite a bit of time.

Example from my notes:

| Last used | Pri0 | Pri1 | Pri2 | Pri3 | Pri4 |
|-----------|------|------|------|------|------|
| 0         | 4    | 0    | 1    | 2    | 3    |
| 1         | 3    | 4    | 0    | 1    | 2    |
| 2         | 2    | 3    | 4    | 0    | 1    |
| 3         | 1    | 2    | 3    | 4    | 5    |
| 4         | 0    | 1    | 2    | 3    | 4    |

devcrafter replied:
Thanks, yes, will add the comment. Your math is correct 💯

devcrafter committed:
Add comment & polishing

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Fix: round-robin load balancing for custom key queries

+ even subqueries distribution over available nodes for custom key query
+ creating load balancing instance for each custom key query separately

devcrafter committed:
Fix test

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=57235) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=57235) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: devcrafter<br/>:x: Ubuntu<br/><hr/>**Ubuntu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=57235) it.</sub>

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

devcrafter committed:
Fix test flakiness with hedge connections enabled

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-custom-key-failover

tavplubix commented:
@devcrafter, the new test is flaky: https://s3.amazonaws.com/clickhouse-test-reports/0/de3e7012d5b3e07fff7c9bc646e0cc688b903f79/integration_tests__tsan__[3_6].html

devcrafter commented:
> @devcrafter, the new test is flaky: [https://s3.amazonaws.com/clickhouse-test-reports/0/de3e7012d5b3e07fff7c9bc646e0cc688b903f79/integration_tests__tsan__[3_6].html](https://s3.amazonaws.com/clickhouse-test-reports/0/de3e7012d5b3e07fff7c9bc646e0cc688b903f79/integration_tests__tsan__%5B3_6%5D.html)

Thanks! Will check it
